Reduce the closure size of various qt6 packages

NickCao opened the pull request:
###### Description of changes
Fixes https://github.com/NixOS/nixpkgs/issues/208246

I no longer see qtbase checking for valgrind in configurePhase, let's just remove it. This should not only reduce the runtime closure size, but also the build time closure size by a huge margin.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

NickCao committed:
qt6.qtvirtualkeyboard: add dev output

NickCao committed:
qt6Packages.quazip: add dev output

NickCao committed:
qt6.qtbase: remove valgrind from propagatedBuildInputs

reduces the build-time closure and the runtime closure of packages accidentally referencing qtbase.dev by nearly 600M

NickCao commented:
Meanwhile our qtdoc is utterly broken, it has been producing empty output, this is a task for another day though.

NickCao commented:
For those new to this issue:

TLDR: there are currently three problems:
- `qtbase.dev` is bloated by propagated `valgrind`
- `out` of `qt6.qt*` without a `dev` output is bloated by propagated `qtbase.dev`
- some packages depending on `qtbase` is bloated by reference to `qtbase.dev` due to `__FILE__` macro in `${qtbase.dev}/include/QtTest/qtestmouse.h`

They are fixed as such:
- removing `valgrind` from the `propagatedBuildInputs` of `qtbase`
- add the missing `dev` output
- `NIX_CFLAGS_COMPILE = [ "-fmacro-prefix-map=${qtbase.dev}=qtbase.dev" ];`

NickCao committed:
openboard: fix build

NickCao committed:
fritzing: fix build

NickCao commented:
Result of `nixpkgs-review pr 208440` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>5 packages marked as broken and skipped:</summary>
  <ul>
    <li>gmic_krita_qt</li>
    <li>ideamaker</li>
    <li>jami-client-qt</li>
    <li>krita-beta</li>
    <li>obs-studio-plugins.obs-hyperion</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>obs-studio-plugins.obs-ndi</li>
  </ul>
</details>
<details>
  <summary>96 packages built:</summary>
  <ul>
    <li>beamerpresenter</li>
    <li>beamerpresenter-poppler</li>
    <li>calcmysky</li>
    <li>calibre</li>
    <li>corectrl</li>
    <li>cudaPackages.nsight_compute</li>
    <li>cutemaze</li>
    <li>duckstation</li>
    <li>fcitx5-chinese-addons</li>
    <li>fcitx5-configtool</li>
    <li>fcitx5-unikey</li>
    <li>fcitx5-with-addons</li>
    <li>fritzing</li>
    <li>gmic-qt-krita</li>
    <li>gpxsee-qt6</li>
    <li>jami-client</li>
    <li>krita</li>
    <li>ladybird</li>
    <li>libsForQt5.fcitx5-qt</li>
    <li>libsForQt5.quazip</li>
    <li>maestral-gui</li>
    <li>mediaelch</li>
    <li>mediaelch-qt6</li>
    <li>mozillavpn</li>
    <li>nomacs</li>
    <li>obs-studio</li>
    <li>obs-studio-plugins.input-overlay</li>
    <li>obs-studio-plugins.looking-glass-obs</li>
    <li>obs-studio-plugins.obs-backgroundremoval</li>
    <li>obs-studio-plugins.obs-gstreamer</li>
    <li>obs-studio-plugins.obs-move-transition</li>
    <li>obs-studio-plugins.obs-multi-rtmp</li>
    <li>obs-studio-plugins.obs-nvfbc</li>
    <li>obs-studio-plugins.obs-pipewire-audio-capture</li>
    <li>obs-studio-plugins.obs-source-record</li>
    <li>obs-studio-plugins.obs-vkcapture</li>
    <li>obs-studio-plugins.wlrobs</li>
    <li>openboard</li>
    <li>pokefinder</li>
    <li>prismlauncher</li>
    <li>prismlauncher-qt5</li>
    <li>python310Packages.pyqt6</li>
    <li>python310Packages.pyqt6-webengine</li>
    <li>python39Packages.pyqt6</li>
    <li>python39Packages.pyqt6-webengine</li>
    <li>qMasterPassword</li>
    <li>qmapshack</li>
    <li>qmmp</li>
    <li>qt6.full</li>
    <li>qt6.qmake</li>
    <li>qt6.qt3d</li>
    <li>qt6.qt5compat</li>
    <li>qt6.qtbase</li>
    <li>qt6.qtcharts</li>
    <li>qt6.qtconnectivity</li>
    <li>qt6.qtdatavis3d</li>
    <li>qt6.qtdeclarative</li>
    <li>qt6.qtdoc</li>
    <li>qt6.qthttpserver</li>
    <li>qt6.qtimageformats</li>
    <li>qt6.qtlanguageserver</li>
    <li>qt6.qtlottie</li>
    <li>qt6.qtmultimedia</li>
    <li>qt6.qtnetworkauth</li>
    <li>qt6.qtpositioning</li>
    <li>qt6.qtquick3d</li>
    <li>qt6.qtquick3dphysics</li>
    <li>qt6.qtquicktimeline</li>
    <li>qt6.qtremoteobjects</li>
    <li>qt6.qtscxml</li>
    <li>qt6.qtsensors</li>
    <li>qt6.qtserialbus</li>
    <li>qt6.qtserialport</li>
    <li>qt6.qtshadertools</li>
    <li>qt6.qtspeech</li>
    <li>qt6.qtsvg</li>
    <li>qt6.qttools</li>
    <li>qt6.qttranslations</li>
    <li>qt6.qtvirtualkeyboard</li>
    <li>qt6.qtwayland</li>
    <li>qt6.qtwebchannel</li>
    <li>qt6.qtwebengine</li>
    <li>qt6.qtwebsockets</li>
    <li>qt6.qtwebview</li>
    <li>qt6Packages.kcoreaddons</li>
    <li>qt6Packages.poppler</li>
    <li>qt6Packages.qtpbfimageplugin</li>
    <li>qt6Packages.quazip</li>
    <li>qt6Packages.qxlsx</li>
    <li>qutebrowser-qt6</li>
    <li>retext</li>
    <li>stellarium</li>
    <li>tdesktop</li>
    <li>teamspeak_client</li>
    <li>zeal-qt6</li>
    <li>zint</li>
  </ul>
</details>

vcunat commented:
In some cases I addressed the `__FILE__` issues (via `assert` or similar) by injecting `#line` directives, e.g. in
https://github.com/NixOS/nixpkgs/blob/2ccc3219c54e3c62c2bb3122b1c651c607029d0b/pkgs/servers/x11/xorg/overrides.nix#L884

NickCao commented:
> In some cases I addressed the `__FILE__` issues (via `assert` or similar) by injecting `#line` directives

This approach is less applicable on qt, which has a bunch of macros like `QTEST_ASSERT`, and their definitions are scattered everywhere.

NickCao committed:
qt6.qtbase: patch QTEST_ASSERT and other macros to remove reference to qtbase.dev via __FILE__ macro

This is safe because they are only used for printing debug messages, alternatives are setting -fmacro-prefix-map=${qtbase.dev}=qtbase.dev in NIX_CFLAGS_COMPILE or injecting #line directives in all header files, but they would interfere with other usages of __FILE__ macro that might depend on the the path to actually point to a file.

NickCao committed:
Revert "qt6.qtdeclarative: reduce closure size by removing reference to qtbase.dev"

This reverts commit cec6862940c2deb71b599536dc31c5d10a43b5d6.

NickCao commented:
Stilled worried about potential issue around mangling with `__FILE__` macros, let's only modify a single header. Despite `__BASE_FILE__` being a GNU C extension, it's well supported by clang and older version of gcc.

Atemu commented on pkgs/development/libraries/qt-6/modules/qtbase.nix:
```diff
@@ -140,7 +139,6 @@ stdenv.mkDerivation rec {
     libthai
     libdrm
     libdatrie
-    valgrind
```
Would be nice to have the explanation for this in the commit message.

Atemu commented on pkgs/development/libraries/qt-6/modules/qtvirtualkeyboard.nix:
```diff
@@ -11,5 +11,4 @@ qtModule {
   qtInputs = [ qtbase qtdeclarative qtsvg ];
   propagatedBuildInputs = [ hunspell ];
   nativeBuildInputs = [ pkg-config ];
-  outputs = [ "out" ];
```
I'm confused, how does this add a dev output?

NickCao replied:
`qtModule` by default sets `output = [ "out" "dev" ];`

Atemu commented:
qt6.qtbase: patch QTEST_ASSERT and other macros to remove reference to qtbase.dev

->

    qt6.qtbase: remove reference to qtbase.dev

    Achieved by patchig QTEST_ASSERT and other macros. This is safe because ... alternatives are ... but ...

fufexan reviewed (approved):
Openboard LGTM.
